### PR TITLE
Commit paths

### PR DESCRIPTION
**Pull request by bpeabody**
allows users to restrict commit changes to specific paths

### CODE REVIEW

**Comment by bpeabody** [comment]
Support for paths with `--amend` is coming.

File: node/lib/util/commit.js [hunk 1]
```diff
@@ -200,6 +200,25 @@ exports.formatEditorPrompt  = function (status, cwd, all) {
 };
 
 /**
+ * Stage all of the specified `submodules` that are open in the specified
+ * `index`.  We need to do this whenever generating a meta-repo commit because
+ * otherwise, we could commit a staged commit in a submodule that would have
+ * been reverted in its open repo.
+ * 
+ * @param {NodeGit.Repository} index
```
Comment (novalis) [inline node/lib/util/commit.js]: That parameter is an index not a repo, right?

File: node/lib/cmd/commit.js [hunk 2]
```diff
@@ -89,27 +89,52 @@ Amend the last commit, including newly staged chnages and, (if -a is \
 specified) modifications.  Will fail unless all submodules changed in HEAD \
 have matching commits and have no new commits.`,
     });
+    parser.addArgument(["file"], {
+        type: "string",
+        help: `\
+When files are given on the command line, the command commits the contents of
+the named files, without recording the changes already staged. The contents of
+these files are also staged for the next commit on top of what have been
```
Comment (novalis) [inline node/lib/cmd/commit.js]: I'm confused by this text

**Comment by bpeabody** [comment]
Addressed issues noted so far.

File: node/lib/cmd/commit.js [hunk 3]
```diff
@@ -270,6 +332,10 @@ const doAmend = co.wrap(function *(args) {
  * @param {String}  [args.message]
  */
 exports.executeableSubcommand = function (args) {
+    if (args.all && 1 !== args.file.length) {
```
Comment (novalis) [inline node/lib/cmd/commit.js]: Here, you check 1 !== args.file.length, but further up, you check 0 == args.file.length.  Which is it?

File: node/lib/util/commit.js [hunk 4]
```diff
@@ -316,6 +318,180 @@ exports.commit = co.wrap(function *(metaRepo, all, metaStatus, message) {
...
+
+    for (let filename in staged) {
+        const stat = staged[filename];
+        if (FILESTATUS.REMOVED === stat) {
+            changes[filename] = null;
+        }
+        else {
+            const filePath = path.join(workdir, filename);
+
+            // This API is unfinished; instead of returning an id, it takes an
```
Comment (novalis) [inline node/lib/util/commit.js]: "this api" meaning createFromDisk?

File: node/lib/util/commit.js [hunk 5]
```diff
@@ -316,6 +318,180 @@ exports.commit = co.wrap(function *(metaRepo, all, metaStatus, message) {
...
+            yield index.addByPath(filename);
+        }
+    }
+
+    // Then submodules.
+
+    const subs = status.submodules;
+    for (let subName in subs) {
+        const sub = subs[subName];
+        if (null !== sub.commit &&
```
Comment (novalis) [inline node/lib/util/commit.js]: what about new subs?

File: node/lib/cmd/commit.js (hunk 3 above)
Comment (bpeabody) [inline node/lib/cmd/commit.js]: yeah, that's wrong

File: node/lib/util/commit.js (hunk 4 above)
Comment (bpeabody) [inline node/lib/util/commit.js]: noted, improved

File: node/lib/util/commit.js (hunk 5 above)
Comment (bpeabody) [inline node/lib/util/commit.js]: clarified

**Comment by bpeabody** [comment]
addressed comments

File: node/lib/util/commit.js [hunk 6]
```diff
@@ -729,3 +905,171 @@ Author:    ${commitSig.name()} <${commitSig.email()}>
     result += exports.formatStatus(status, cwd, all);
     return "\n" + prefixWithPound(result) + "\n";
 };
+
+/**
+ * Calculate and return, from the specified `current` `RepoStatus` object
+ * describing the statue of the repository, and the specified `requested`
```
Comment (novalis) [inline node/lib/util/commit.js]: s/statue/state/

**Comment by bpeabody** [comment]
Not sure what my deal is with statues, but fixed that last one.

**Pull request merged**